clar2wasm/src/standard/standard.wat

```diff
@@ -2598,42 +2761,59 @@
 
         ;; Copying the data from the offset to isolated environment (i.e. target-index = $stack-pointer+(initial-values+(80 rounds * 8)))
+        ;; Leaving some space to store 64 words, then storing actual data
+        ;; 80 words of 80 bytes = 640 + 64 (bytes for intial values) = 704
         (memory.copy (i32.add (global.get $stack-pointer) (i32.const 704)) (local.get $offset) (local.get $length))
 
+        ;; Making sure that $res_len is divisible by 128 ($res_len%128==0)
         (local.set $length_after_padding ;; total size of data with expansion divisible by 128
             (i32.add
                 (i32.or
                     ;; len + 16 bytes for the size
+                    ;; Note that even if we're handling length in 8 bytes, but it's important that we calculate final length, 
+                    ;; considering that length will be stored in 16 bytes. For example if we consider length here as 8 bytes instead 
+                    ;; of 16, the input with length of 119 would cause the problem. SHA-512 should create another block if length of
+                    ;; data to be hashed is greater than 111, because it expects last 16 bytes to be length, and there is one byte for
+                    ;; '1' added after data. 111+1+16 is exactly 128. If we don't put 16 here, it will consider length to be 8 byte,
+                    ;; and won't create another block even if data length is 119. If we're not calculating the length in 16 bytes, it
+                    ;; doesn't mean that we should put data in those 8 bytes we're ignoring. It should be filled with zeros instead
+
+                    ;; $length + 16 bytes for the size
                     (i32.add (local.get $length) (i32.const 16))
+                    ;; 0x7f=127=1111111 to make it divisible by 128
                     (i32.const 0x7f)
                 )
+                ;; 1 byte for '1' to add at the end of the data
                 (i32.const 1)
             )
         )
 
         ;; Add "1" at the end of the data
         (i32.store offset=704
             (i32.add (global.get $stack-pointer) (local.get $length))
+            ;; 0x80=128=10000000
             (i32.const 0x80)
         )
 
         ;; Fill the remaining part before the size with 0s
         (memory.fill
+                ;; 704 + 1 ('1' added after the data) = 705
                 (i32.add (i32.add (global.get $stack-pointer) (local.get $length)) (i32.const 705))
                 (i32.const 0)
                 ;; Leave the last 8 bytes for the length
-                ;; Not handling 16 bytes length here
+                ;; We're subtracting 8 because all other bytes will be filled with zeros, including ignored 8 bytes from 16 byte
+                ;; length
                 (i32.sub (i32.sub (local.get $length_after_padding) (local.get $length)) (i32.const 8))
         )
 
-        ;; Add the length, as a 64bits big-endian integer
+        ;; Convert length to bits, (i.e. 13 bytes = 104 bits)
```

Comment:
Still believe the previous comment was more explicit

Reply:
Improved